test/test_strdup.cpp

```diff
@@ -73,3 +73,18 @@ TEST(test_strdup, invalid_arguments) {
   EXPECT_EQ(NULL, rcutils_strndup(NULL, 5, allocator));
   EXPECT_EQ(NULL, rcutils_strdup("something", failing_allocator));
 }
+
+TEST(test_strndup, one_byte_overread) {
+  auto allocator = rcutils_get_default_allocator();
+  char str[4];
+  char * p;
+  memcpy(str, "test", sizeof(str));
+
+  // If there is a bug, a one byte overread happens here. Run this test under a
+  // memory sanitizer to guarantee it causes a crash.
+  p = rcutils_strndup(str, sizeof(str), allocator);
+  if (NULL == p) {
+    FAIL();
+  }
+  allocator.deallocate(p, allocator.state);
```

Comment:
```suggestion
  ASSERT_STREQ(str, "test");
  allocator.deallocate(p, allocator.state);
```

That is, let's make sure the string contains what we expect.

Reply:
There's a bug in the suggestion: a string comparison between `str` and `"test"` could try to read a 5th byte in `str`. Do you mean to compare `p` and `"test"` instead?